README
POE::Component::Generic is POE component that provides non-blocking access
to a blocking object. 

It works by forking of a sub-process with POE::Wheel::Run to do the work
The object is created in the child process, which waits for method calls 
from the parent.

It is intended for people who want to use an object that blocks from
POE with as little pain as possible.


NOTE: It has come to our attention that POE::Component::Generic does not
always function correctly on Windows.  We do not have the skills nor the
time to debug this issue.  We would be very pleased if you would solve this
for us.


POE::Component::Generic can be installed the usual ways:

Through the CPAN shell:

	perl -MCPAN -e 'install "POE::Component::Generic"'

Or by downloading it and running the usual commands:

	perl Makefile.PL
	make test
	make install

It is possible to configure some of the test cases by passing arguments to
Makefile.PL.

    perl Makefile.PL --help


The project needs your help.  If you're interested, please contact
Philip Gwyn at gwyn-at-cpan.org.

Thank you.




Changes
Revision history for POE::Component::Generic.

0.1402 Nov 28 2011
    - Use FindBin to get t:: into @INC
    - Allow test delay to be long to deal with slow smokers

0.1401 Aug 26 2011
    - Added t/00_info.t

0.1400 Wed May 18 2011
    - Added ->kill()
    - Added on_exit

0.1300 Wed May 18 2011
    - Bump copyright date
    - Don't blindly call $obj->DESTROY, as per rt68062
    - Confirmed to work with POE 1.311
    - Tentatively removed the limit of alt_fork on Win32, as per rt67072

0.1205 Wed Sep 16 2009
    - Tweak 02_pod_coverage.t
    - Require 5.6
    - no strict 'subs'; in Net::SSH2 code

0.1204 Tue Sep 15 2009
    - 21_alt_fork-x.t blindly chmod, then check for +x

0.1203 Tue Sep 15 2009

    - 21_alt_fork-x.t checks to make sure t/alt-fork is +x.  If not, skip
      tests.  5.10.0 on Ubuntu wasn't setting +x on anything in t/
        Thank you Naveed Massjouni.

0.1202 Fri Sep 11 2009
    - Use POE::Wheel::Run->PID instead of $$ from the child

0.1201 Wed May 06 2009
    - Fix a bunch of warnings that were previously hidden by POE turning off
        warnings.
        - Thank you Hinrik
    - Added 01_pod and 02_pod_coverage

0.1200 Wed Mar 11 2009
    - Better handling of shutdown, delete wheel on CHLD.  Now the session
      may _stop, even if the kernel keeps on.
        Thank you Paul Williams.  Closes RT 44039
    - Unit tests for above
    - Create a smaller closure to run the child process
    - Tweak debug messages
    - Tweak doco intro

0.1101 Thu Sep 04 2008 
    - Internal release

0.1100 Wed Jan 23 2008
    - Did away with mapping PBids to session/event tupples.  This allows
        us to completely ignore postback GC issues, at the cost of slightly
        modifying postback semantics.  Win!
        Closes RT 32067, thank you Paul Driver.

0.1008 Thu Nov 29 2007
    - t/91_ssh.t wasn't skipping out early if Net::SSH2 wasn't installed

0.1007 Thu Nov 29 2007
    - Fixed POD synopsis of PoCo::Generic::Net::SSH2
    - Fixed PoCo::Generic::Net::SSH2 to work with Net::SSH2 0.18
    - Fixed error about package methods
    - Improved explanations in the POD.
    - Added examples.
    - Don't allow 0-9 in object ID

0.1006 Fri Aug 17 2007
    - Fixed t/21-alt_fork for NetBSD.  
        Thank you BinGOs

0.1005 Wed Aug 15 2007
    - t/alt-fork uses /bin/sh

0.1004 Wed Aug 15 2007
    - Make sure we have /usr/bin/perl

0.1003 Tue Aug 14 2007
    - Don't use Data::Denter !
    - Don't expect @INC to be a given length !

0.1002 Tue Aug 14 2007
    - alt_fork can be a path to the perl executable
        Thank you Matt Sickler for the idea

0.1001 Thu Jun 7 2007
    - refcount_increment on the dest session, not $sender
        Thank you David Snopek

0.1000
    - Added {error}

0.0911 Mon Dec 4 2006
    - Moved to subversion.  Removed Log tags.
    - Versions in PM_PREREQ now numbers

0.0910 Wed Nov 15 2006
    - Warnings about alt_fork on MSWin32.
    - Skip 20_alt_fork.t.
    - Don't use alt_fork in the other test cases.

0.0905 Tue Aug 29 2006
    - Better example in PoCo::Generic::Object about factories.

0.0904 Wed May 3 2006
    - Forgot to add t/P15.pm to manifest

0.0903 Tue May 2 2006
    - In child, dup STDOUT to talk to parent and dup STDERR to STDOUT so
      that user code can't mess up the output.
            Thanks to Jeremy Palmer for pushing me into this idea.

0.0901 Tue Apr 18 2006
    - Fix alt_fork for MSWin32.  Reported working by BinGOs, thank you.

0.09 Tue Apr 18 2006
    - Fix Ext::MM::Coverage->ignore in Makefile.PL
    - Use Win32::LoginName on MSWin32 in Makefile.PL
    - Tests now check delay within one second

0.08 Thu Apr 13 2006
    - New vs spawn in the examples
    - Work around in PoCo::Generic::Net::SSH2 so that the PAUSE indexer 
        doesn't complain
    - PoCo::Generic::Net::SSH2 now works with alt_fork=>1

0.07 Wed Apr 12 2006
    - Added documentation
    - Added __callback_argument and __postback_arguement
    - Use Scalar::Util::reftype instead of ref()
    - Added __package_register
    - Added PoCo::Generic::Net::SSH2->exec and ->cmd
    - Fixed PoCo::Generic::Object->DESTROY
    - Added test cases to improve test coverage

0.06 Tue Apr 11 2006
    - Added PoCo::Generic::Net::SSH2
    - Added t/91_ssh2.t which tests the above
    - Added t/90_ssh.t which tests Net::SSH::Perl
    - Added options to Makefile.PL to config the above 2 tests
    - Added Generic->package_register
    - Added Generic->new options :
            - packages to configure factory generated packages
            - child_package so we can overload the Child process behaviour
    - Object IDs now get incremented properly
    - Child now uses get_requests to fetch the next requests

0.05 Mon Apr 10 2006
    - Major refactoring
    - Added factory methods
    - Split spawn() and new() apart

0.04 Thu Apr 6 2006
    - Added callbacks
    - Renamed all the wheel_* events to __wheel_* so that the only
      method name that can't be called on an object is 'shutdown'.

0.03 Thu Apr 6 2006
    - Added callbacks -- still need postbacks

0.02 Thu Apr 6 2006
    - Brown paper bag edition
    - Removed Data::Denter usage from test cases
    - Updated README

0.01 Wed Apr 5 2006
    - first version based on my module poco client daap


t/P4.pm
# $Id: P4.pm 109 2006-04-07 20:15:31Z fil $
package t::P4;
use strict;

use vars qw( @ISA );

@ISA = 'P1';

sub in4 {}

1;
__END__

$Log$
Revision 1.1  2006/04/07 20:15:31  fil
Initial revision


t/P10.pm
# $Id: P10.pm 117 2006-04-12 08:13:22Z fil $
package t::P10;
use strict;

sub DEBUG () { 0 }

sub new
{
    my( $package, %args ) = @_;
    DEBUG and warn "new";
    return bless { %args }, $package;
}

sub delay
{
    my( $self ) = @_;
    DEBUG and warn "$self->delay";
    my $before=time;
    sleep( $self->{delay} );
    DEBUG and warn "AFTER";
    return ($before, time);
}

sub set_delay
{
    my( $self, $new ) = @_;
    DEBUG and warn "$self->set_delay( $new )";
    $self->{delay} = $new;
    return;
}

sub get_delay
{
    my( $self ) = @_;
    return $self->{delay};
}

sub die_for_your_country
{
    my( $self, $text ) = @_;
    die $text;
}

1;
__END__

$Log$
Revision 1.2  2006/04/12 08:13:22  fil
Added documentation
Added __callback_argument and __postback_arguement
Use Scalar::Util::reftype instead of ref()
Added __package_register
Added PoCo::Generic::Net::SSH2->exec and ->cmd
Fixed PoCo::Generic::Object->DESTROY
Added test cases to improve test coverage

Revision 1.1.1.1  2006/04/07 20:15:31  fil
Log



t/P30.pm
# $Id: P30.pm 109 2006-04-07 20:15:31Z fil $
package t::P30;
use strict;

sub DEBUG () { 0 }

sub new
{
    my( $package, %args ) = @_;
    DEBUG and warn "new";
    return bless { %args }, $package;
}

sub something
{
    my( $self, $one, $coderef, $two ) = @_;
    
    my $answer = $one + $two;
    $coderef->();
    return $answer;
}

sub otherthing
{
    die "Not implemented";
}

sub twothing
{
    my( $self, @coderef ) = @_;
    
    for ( 1, 2 ) {
        foreach my $code ( @coderef ) {
            $code->( $_ );
        }
    }
    return;
}




1;
__END__

$Log$
Revision 1.1  2006/04/07 20:15:31  fil
Initial revision



Makefile.PL
#!/usr/bin/perl

use warnings;
use strict;

use ExtUtils::MakeMaker;
use Getopt::Long;
use Data::Dumper;

use 5.006;

# Add in the Ex::MM::Coverage stuff.  But silently ignore failure.
eval q{
    use ExtUtils::MakeMaker::Coverage;
    # I keep some modules outside of /usr/lib/perl5
    my $config = ExtUtils::MakeMaker::Coverage->config;
    $config->ignore( ['site_perl'] ); 
};


###########################################################
my $HELP = 0;
my %config = (
    ssh=>{
        host => 'localhost',
        port => 22,
        user => '',
        password => '',
    }
);
if( $^O =~ /MSWin32/i or $^O =~ /cygwin/ ) {
    warn <<'    WARN';
************************************************************
It has come to our attention that POE::Component::Generic
does not always function correctly on Windows.  We do not
have the skills nor the time to debug this issue.  We would
be very pleased if you would solve this for us.  
    Thank you.
************************************************************
    WARN
    eval q{
        use Win32;  
        $config{ssh}{user} = Win32::LoginName() unless Win32::IsWin95();
    };
}
else {
    $config{ssh}{user} = getpwuid($<);
}


my @args = @ARGV;
my $res = GetOptions ( "ssh-host=s" => \$config{ssh}{host}, 
                       "ssh-port=i" => \$config{ssh}{port}, 
                       "ssh-user=s" => \$config{ssh}{user}, 
                       "ssh-password=s" => \$config{ssh}{password},

                       "help"       => \$HELP
                     );

if( not $res or $HELP ) {
    print <<HELP;
Usage: $0 [options] [MakeMaker-optoins]

    --help          This text

For testing Net::SSH::Perl
    --ssh-host      Host to connect to ($config{ssh}{host})
    --ssh-port      Port to connect to ($config{ssh}{port})
    --ssh-user      User to login as ($config{ssh}{user})
    --ssh-password  Password to use
HELP
}

# @ARGV=@args;

###########################################################
open CONFIG, ">t/Config.pm" or die "Unable to create t/Config.pm: $!";
my $conf = Dumper \%config;
print CONFIG <<CONFIG;
package t::Config;
use vars qw( \$VAR1 );
$conf
1;
CONFIG

close CONFIG;



###########################################################

WriteMakefile(
	NAME          => 'POE::Component::Generic',
	AUTHOR        => 'Philip Gwyn <gwyn-at-cpan.org>',
	ABSTRACT      => 'Generic non-blocking POE interface to any OO-module',
	VERSION_FROM  => 'lib/POE/Component/Generic.pm', # finds $VERSION
    LICENSE       => 'perl',
	PREREQ_PM     => {
		'Test::More'        => 0.32,
		'POE'               => 1.311,
        'Scalar::Util'      => 1,
		'Devel::Symdump'	=> 2,
		'Carp'			    => 0,
	},
	dist        => {
		COMPRESS  => 'gzip -9f',
		SUFFIX    => 'gz',
	},
);




example/demo.pl
#!/usr/bin/perl -w

use strict;

use POE;
use POE::Component::Generic;

my $obj = POE::Component::Generic->new(
	package => 'My::Package',
        alias   => 'my-object',
        debug	=> 0,
        object_options => [ ten => 10, answer=>42 ]
    );         

POE::Session->create(
    inline_states => {
      _start => sub {
          $poe_kernel->delay( 'first', 1);	# give child some time
      },

      # Almost direct object method
      first => sub {
          $obj->fetch( {event=>'got_ten'}, 'ten' );
          return;
      },
      
      # POE-style post to component
      got_ten => sub {
          my( $data, $ten ) = @_[ARG0, ARG1];

          print "$ten == 10\n";
          print "$data->{result}[0] == 10\n";
          
          $poe_kernel->post( 'my-object',
                             fetch => {event=>'got_answer'}, 'answer' );
      },
      
      # ->yield to the component object, showing off wantarray
      got_answer => sub {
          my( $data, $answer ) = @_[ARG0, ARG1];

          print "$answer == 42\n";
          print "$data->{result}[0] == 42\n";
          
          $obj->yield( all_keys => {event=>'the_keys', wantarray=>1} );
      },
      
      # ->call to the component object
      the_keys => sub {
          my( $data, @keys ) = @_[ ARG0, ARG1..$#_ ];
          
          print join ', ', @keys;
          print " == qw( ten answer )\n";
          
          $obj->call( badness => {event=>'yow'} );
      },
      
      # Show error reporting
      yow => sub {
          my( $data ) = @_[ ARG0 ];

          # $data->{result} eq undef
          print qq($data->{error} eq "KABLOUIE! at line ..."\n);

          $poe_kernel->post( 'my-object' => 'shutdown' );
      }
    },
    
);


$poe_kernel->run;
    


#############################################################
package My::Package;

use strict;


sub new
{
    my $package=shift;
    return bless { @_ }, $package;
}

sub fetch
{
    my( $self, $key ) = @_;
    warn "Returning $key";
    return $self->{$key};
}

sub all_keys
{
    my( $self ) = @_;
    return keys %$self;
}

sub badness
{
    die "KABLOUIE!";
}


example/postback.pl
#!/usr/bin/perl -w

use strict;

$|++;

# sub POE::Kernel::TRACE_DEFAULT () { 1 }

use POE;
use POE::Component::Generic;

my $obj = POE::Component::Generic->new(
	package => 'My::Package',
        alias   => 'my-object',
        debug	=> 1,
        verbose => 1,
        object_options => [ wait => 5, answer=>42 ],
                        # first argument of ->setup is a postback
        postbacks => { setup=>0 }
        
    );         

POE::Session->create(
    inline_states => {
        _start => sub {
            $poe_kernel->delay( 'first', 1);	# give child some time
        },

        # 'answer' is a postback
        first => sub {
            print "Setting a postback\n";
            $obj->setup( { event=>'setup_done' }, 'got_answer' );
        
            return;
        },
      
        setup_done => sub {
            print "Calling the postback\n";
            # This will cause the object to use our postback
            $obj->doit( {} );
        },
      
        # This is the postback.  Notice that ARG0 isn't the data hash
        # like a response.
        got_answer => sub {
            my( $answer ) = $_[ ARG0 ];

            print "$answer == 42\n";

            $obj->shutdown;          
        },
    }
);


$poe_kernel->run;
    


#############################################################
package My::Package;

use strict;


sub new
{
    my $package=shift;
    return bless { @_ }, $package;
}

sub setup
{
    my( $self, $coderef ) = @_;
    $self->{coderef} = $coderef;
    print "Coderef set\n";
}

sub doit
{
    my( $self ) = @_;
    print "Sleeping $self->{wait} seconds\n";
    sleep( $self->{wait} );
    print "Calling coderf\n";
    $self->{coderef}->( $self->{answer} );
    return;
}



lib/POE/Component/Generic/Net/SSH2.pm
package POE::Component::Generic::Net::SSH2;
# $Id: SSH2.pm 759 2011-05-18 16:55:01Z fil $

use strict;

use POE::Component::Generic;
use POE::Component::Generic::Child;

use Carp qw(carp croak);
use vars qw( @ISA $TIMEOUT $VERSION );

use         # hide from CPANTS
    Net::SSH2;

$VERSION = '0.1300';
@ISA = qw( POE::Component::Generic );
$TIMEOUT = 100;

require Net::SSH2;
if( $Net::SSH2::VERSION < 0.18 ) {
    croak __PACKAGE__, " requires version Net::SSH2 0.18 or better";
}

##########################################################
#
# Create the worker object
# This is called from PoCo::Generic->spawn()
# We set-up everything for the package, so the user doesn't have to.
sub new
{
    my( $package, %params ) = @_;

    $params{package}    ||= 'Net::SSH2';
    $params{factories}  ||= [ qw( channel real_exec ) ];
    $params{packages}   ||= {};
    $params{packages}{'Net::SSH2::Channel'}{ postbacks }
                    ||= [ qw(handler_stderr handler_stdout handler_closed) ];
    $params{postbacks}  ||= { disconnect=>0, real_exec=>[1..4] };
    $params{alias}      ||= 'net-ssh';
    $params{child_package} ||= join '::', __PACKAGE__, 'Child';

    $TIMEOUT = delete $params{timeout} if $params{timeout};

    $package->SUPER::new( %params );
}


##########################################################
# This method converts the SomethingEvent hash into a linear
# argument list, so that the user doesn't have to.
sub exec
{
    my( $self, $data, $command, %events ) = @_;

    my @args = ( $data, $command );
    foreach my $ev ( qw(Stdout Stderr Closed Error) ) {
        push @args, $self->__postback_argument( $ev, \%events );
    }

    $self->real_exec( @args );
}

#####################################################################
# Child process worker
package POE::Component::Generic::Net::SSH2::Child;
use strict;

use IO::Poll qw(POLLRDNORM POLLIN);
use Scalar::Util qw( reftype blessed );
use vars qw( @ISA );

sub DEBUG () { 0 }

@ISA = qw( POE::Component::Generic::Child );


sub new
{
    my $package = shift;
    my $self = $package->SUPER::new( @_ );

    $self->{timeout} ||= $POE::Component::Generic::Net::SSH2::TIMEOUT;
    
    $Net::SSH2::CHILD = $self;
    return $self;
}

#######################################
# Poll stdin and the SSH connection
sub get_requests
{
    my( $self ) = @_;

    unless( $self->{ssh_channels} ) {       # no registered SSH channels
        return $self->SUPER::get_requests();
    }
    DEBUG and warn "we have ssh_channel handlers";

    my $rv;
    while( 1 ) {
        $self->poll_ssh;                    # let SSH channels have a chance

        $rv = $self->poll_stdin;
        return $rv if defined $rv;
    }
}

#######################################
sub poll_ssh
{
    my( $self ) = @_;

    my @poll;
    while( my( $name, $hdef ) = each %{ $self->{ssh_channels} || {} } ) {
        push @poll, { handle=>$hdef->{channel},
                      events=>[ grep { $hdef->{$_} } qw( in ext ) ]
                    };
    }
    return unless @poll;
    DEBUG and warn "ssh_poll";
    # Blocking has to be off for polling
    $self->{obj}->blocking( 0 );
    $self->{obj}->poll( $self->{timeout}, [ @poll ] );

    DEBUG and warn "ssh_poll returned ", 0+@poll, " events";

    my( $n, $buf );
    foreach my $poll ( @poll ) {
        my $hdef = $self->{ssh_channels}{ $poll->{handle} };
        foreach my $ev ( qw( in ext ) ) {
            next unless $poll->{revents}{$ev};

            if( $n = $hdef->{channel}->read( $buf, 1024, $ev eq 'ext' ) ) {
                DEBUG and warn "$ev: $n bytes";
                $hdef->{$ev}->( $buf, $n );
            }
        }
        if( $hdef->{channel}->eof ) {
            DEBUG and warn "EOF";
            $hdef->{channel_closed}->();
            delete $self->{ssh_channels}{ $poll->{handle} };
        }
    }
    return;
}

#######################################
sub __handler
{
    my( $self, $channel, $event, $coderef ) = @_;

    $self->{ssh_channels} ||= {};

    if( $coderef ) {
        die "Must be a coderef" unless 'CODE' eq reftype $coderef;
        $self->{ssh_channels}{$channel} ||= { channel=> $channel };
        $self->{ssh_channels}{$channel}{$event} = $coderef;
    }
    else {
        delete $self->{ssh_channels}{$channel}{$event};
        delete $self->{ssh_channels}{$channel} 
                             if 1 == keys %{ $self->{ssh_channels}{$channel} };
        delete $self->{ssh_channels} if 0 == keys %{ $self->{ssh_channels} };
    }
}

#######################################
sub __remove_channel
{
    my( $self, $channel ) = @_;

    return unless $self->{ssh_channels};

    delete $self->{ssh_channels}{$channel};
    delete $self->{ssh_channels} if 0 == keys %{ $self->{ssh_channels} };
    return;
}


#######################################
# Poll STDIN for a request
# Return 
#   undef() when STDIN closed (parent shutdown)
#   0 nothing to do
#   [ requests...] what to do
sub poll_stdin
{
    my( $self ) = @_;

    DEBUG and warn "poll_stdin ($self->{timeout})";
    my $poll = IO::Poll->new();

    $poll->mask( *STDIN, POLLIN );
    $poll->poll( $self->{timeout}/1000 );
    my $raw;
    my $ev = $poll->events( *STDIN );
    DEBUG and warn "ev=$ev";
    return unless $ev;

    # This should be encapsulated in Generic::Child somehow.
    return 0 unless sysread ( STDIN, $raw, $self->{size} );
    return $self->{filter}->get([$raw]);
}






#####################################################################
# Extend Net::SSH2
#   Net::SSH2 doesn't play nicely with subclassing, so we just throw
#   Things into their namespace.
package             # On 2 lines so the PAUSE indexer doesn't gripe
    Net::SSH2;

use strict;
use vars qw( $CHILD );
sub DEBUG () { 0 }

*true_channel = \&channel;

#################################
{
    no strict 'subs';
    *channel = sub {
        my( $self, @args ) = @_;
        # Blocking has to be on when setting up a channel
        # And for ->exec and ->cmd, it seems, but ...
        $self->blocking( 1 );
        my $obj = true_channel( $self, @args );
        warn "Can't create channel ".$self->error unless $obj;
        return $obj;
    };
}

#################################
# This method does the heavy-lifting for ->exec
sub real_exec
{
    my( $self, $command, $on_stdout, $on_stderr, 
                         $on_closed, $on_error ) = @_;

    my $channel = $self->channel;
    unless( $channel ) {
        $on_error->( $self->error ) if $on_error;
        return
    }
    DEBUG and warn "Created channel";

    $CHILD->__handler( $channel, 'in', $on_stdout ) if $on_stdout;
    $CHILD->__handler( $channel, 'ext', $on_stderr ) if $on_stderr;
    $CHILD->__handler( $channel, 'channel_closed', $on_closed ) if $on_closed;
    # $CHILD->__handler( $channel, 'channel_error', $on_error ) if $on_error;

    DEBUG and 
        warn "Exec $command";
    my $ok = $channel->exec( $command );

    if( $ok ) {
        return $channel; 
    }

    $CHILD->__remove_channel( $channel );
    warn "Couldn't exec $command";
    $on_error->( $self->error ) if $on_error;
    return;
}


sub cmd
{
    my( $self, $command, $input ) = @_;

    my $channel = $self->channel;
    unless( $channel ) {
        return ( '', "Unable to create channel: ".$self->error );
    }

    return $channel->cmd( $command, $input );
}


#####################################################################
# Extend Net::SSH2::Channel
#   Net::SSH2 doesn't play nicely with subclassing, so we just throw
#   things into their namespace.
package                     # on 2 lines so PAUSE indexer doesn't gripe
    Net::SSH2::Channel;

use strict;
sub DEBUG () { 0 }

sub handler_stdout
{
    my( $self, $coderef ) = @_;

    DEBUG and warn "handler_stdout";
    $Net::SSH2::CHILD->__handler( $self, 'in', $coderef );
}

sub handler_stderr
{
    my( $self, $coderef ) = @_;

    DEBUG and warn "handler_stderr";
    $Net::SSH2::CHILD->__handler( $self, 'ext', $coderef );
}

sub handler_closed
{
    my( $self, $coderef ) = @_;

    DEBUG and warn "handler_closed";
    $Net::SSH2::CHILD->__handler( $self, 'channel_closed', $coderef );
}

sub cmd
{
    my( $self, $command ) = @_;

    DEBUG and warn "cmd: $command";

    return unless $self->exec( $command );

    my @ret = ( '', '' );

    $self->blocking( 0 );
    my @poll = ({ handle => $self, events => [ qw( in ext ) ] });
    do {
        my( $buf, $n );
        $self->session->poll(100, \@poll);
        if( $self->eof ) {
            DEBUG and 
                warn "EOF";
            return @ret if wantarray;
            return $ret[0];
        }
        foreach my $ev ( qw( in ext ) ) {
            next unless $poll[0]->{revents}->{$ev};

            DEBUG and warn "cmd: read $ev";
            if( $n = $self->read($buf, 1024, $ev eq 'ext') ) {
                DEBUG and warn "$ev: $n bytes";
                $ret[ $ev eq 'in' ? 0 : 1 ] .= $buf;
            }
        }
    } while( 1 );
    # we'd never get here
}

no strict 'subs';
*DESTROY = sub {
        my( $self ) = @_;
        $Net::SSH2::CHILD->__remove_channel( $self ) if $Net::SSH2::CHILD;
    };

1;

__END__

=head1 NAME

POE::Component::Generic::Net::SSH2 - A POE component that provides non-blocking access to Net::SSH2

=head1 SYNOPSIS

    use POE::Component::Generic::Net::SSH2;

    my $ssh = POE::Component::Generic::Net::SSH2->spawn( 
                    alias   => 'my-ssh',
                    verbose => 1,
                    debug   => 0 );

    my $channel;

    POE::Session->create( 
        inline_states => {
            _start => sub {
                $poe_kernel->delay( 'connect', $N );
            },

            connect => sub {
                $ssh->connect( {event=>'connected'}, $HOST, $PORT );
            },
            connected => sub {
                $ssh->auth_password( {event=>'login'}, $USER, $PASSWORD );
            },
            error => sub {
                my( $resp, $code, $name, $error ) = @_[ARG0, $#_];
                die "Error $name ($code) $error";
            },

            login => sub {
                my( $resp, $OK ) = @_[ARG0..$#_];
                unless( $OK ) {
                    $ssh->error( {event=>'error', wantarray=>1} );
                    return;
                }

                $poe_kernel->yield( 'cmd_do' );
            },

            ################
            cmd_do => sub {
                $ssh->cmd( { event=>'cmd_output', wantarray=>1 }, "ls -l" );
                return;
            },
            cmd_output => sub {
                my( $resp, $stdout, $stderr ) = @_[ARG0..$#_];
                warn "Contents of home directory on $HOST:\n$stdout\n";


                $poe_kernel->yield( 'exec_do' );
            },

            exec_do => sub {
                 
                $ssh->exec( { event=>'exec_running', wantarray=>0 }, 
                                "cat - >$FILE", 
                                StdoutEvent => 'exec_stdout', 
                                StderrEvent => 'exec_stderr', 
                                ClosedEvent => 'exec_closed', 
                                ErrorEvent  => 'exec_error' 
                          );
            },
            exec_running => sub {
                my( $resp, $ch ) = @_[ARG0..$#_];
                # keep channel alive
                $channel = $ch;
                $channel->write( {}, "$$-CONTENTS OF THE FILE\n" );
                $channel->send_eof( {event=>'done'} );
            },

            done => sub {
                undef( $channel );
                $ssh->shutdown;
            }

            exec_error => sub {
                my( $code, $name, $string ) = @_[ARG0..$#_];
                die "ERROR: $name $string";
            },        
            exec_stderr => sub {
                my( $text, $bytes ) = @_[ARG0..$#_];
                die "STDERR: $text";
                return;
            },
            exec_stdout => sub {
                my( $text, $bytes ) = @_[ARG0..$#_];
                warn "STDOUT: $text";
                return;
            },
            exec_closed => sub {
                undef( $channel );
                $ssh->shutdown;
            },
        }
    );


=head1 DESCRIPTION


L<POE::Component::Generic::Net::SSH2> is a component for handling SSH2
connections from POE.  It uses L<POE::Component::Generic> to wrap
L<Net::SSH2> into a non-blocking framework.

This component demonstrates many tricks that you might find useful when you
build your own components based on L<POE::Component::Generic>.

It is still ALPHA quality.  Missing are scp, sftp support and better error
handling.

Patches welcome.


=head1 METHODS

=head2 spawn


=head1 Net::SSH2 METHODS

L<POE::Component::Generic::Net::SSH2> supports most Net::SSH2 method calls 
using L<POE::Component::Generic>'s interface.  The following additional 
methods are added for better POE support.


=head2 cmd

    $ssh->cmd( $data_hash, $command, $input );

Ultra-simple command execution.  Runs C<$command> in a new channel on the
remote host.  C<$input> is then feed to it (NOT YET!).  All output is 
accumulated until the command exits, then is sent to the response event
as C<ARG1> (STDOUT) and C<ARG2> (STDERR).


=head2 exec

    $ssh->exec( $data_hash, $command, %events );

Runs C<$command> in a new channel on the remote host.  The response event
will receive an SSH channel that it may use to write to or read from.

C<%events> is a hash that may contain the following keys:

=over 4

=item StdoutEvent

=item StderrEvent

Called when C<$command> writes to STDOUT or STDERR, respectively, with 2
arguments: C<ARG0> is the data, C<ARG1> is the number of bytes.

=item ErrorEvent

Called when there is an SSH error.  The documentation for libssh2 is
piss-poor, so I'm not really sure what these could be, nor how to detect
them all.  Arguments are the same as C<Net::SSH2/error>: C<ARG0> is error
number, C<ARG1> is error name and C<ARG2> is the error string.

=item ClosedEvent

Called when the we encounter an SSH C<eof> on the channel, which normaly
corresponds to when C<$command> exits.

No arguments.

=back


=head1 Net::SSH2::Channel METHODS

A channel is a conduit by which SSH communicates to a sub-process on the
remote side.  Each command, shell or sub-system uses its own channel.  A
channel may only run one command, shell or sub-system.

Channels are created with the C<channel> factory method:

    $ssh->channel( {event=>'got_channel'} );

    # Your got_channel event
    sub got_channel {
        my( $heap, $resp, $channel ) = @_[HEAP, ARG0, ARG1];
        die $resp->{error} if $resp->{error};

        $heap->{channel} = $channel;
    }

You may call most L<Net::SSH2::Channel> methods on a channel using the
normal L<POE::Component::Generic> calling conventions.

    $heap->{channel}->write( {}, $string );

Channels are closed when you drop all references to the object.

    delete $heap->{channel};


There are some extensions to the channel interface:

=head2 cmd

    $heap->{channel}->cmd( {event=>'cmd_response'}, $command, $input );

Runs C<$command> on the channel.  Response event receives 2 arguments:
C<ARG1> is the commands output to STDOUT,
C<ARG2> is the commands output to STDERR.

If you do not set C<wantarray> to 1, you will only receive STDOUT.

=head2 handler_stdout

=head2 handler_stderr

Registers a postback that is posted when the data is present on STDOUT
(called 'in' in libssh2) or STDERR (called 'ext' in libssh2) respectively.  

These could be used when you call L<Net::SSH2::Channel/exec> on a channel.

=head2 handler_closed

Registers a postback that is posted when the channel closes, which normaly
happens when the command has finished.

These could be used when you call L<Net::SSH2::Channel/exec> on a channel.



=head1 AUTHOR

Philip Gwyn E<lt>gwyn-at-cpan.orgE<gt>

=head1 SEE ALSO

L<POE>, L<Net::SSH2>, L<POE::Component::Generic>.

=head1 RATING

Please rate this module.
L<http://cpanratings.perl.org/rate/?distribution=POE-Component-Generic>

=head1 BUGS

Probably.  Report them here:
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=POE%3A%3AComponent%3A%3AGeneric>

=head1 COPYRIGHT AND LICENSE

Copyright 2006, 2011 by Philip Gwyn.


This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
